src/trace_ng/log.zig

```diff
@@ -0,0 +1,672 @@
+        recycle_fba: RecycleFBA,
+        max_buffer: u64,
+        channel: *Channel(logfmt.LogMsg),
+        handle: ?std.Thread,
+
+        pub fn init(config: Config) Self {
+            return .{
+                .allocator = config.allocator,
+                .recycle_fba = RecycleFBA.init(config.allocator, config.max_buffer) catch @panic("could not create RecycleFBA"),
+                .max_buffer = config.max_buffer,
+                .max_level = config.max_level,
+                .exit_sig = config.exit_sig,
+                .channel = Channel(logfmt.LogMsg).init(config.allocator, INITIAL_LOG_CHANNEL_SIZE),
+                .handle = null,
+            };
+        }
+
+        fn unscoped(self: *Self) *UnscopedLogger {
+            return @ptrCast(self);
+        }
+
+        fn withScope(self: *Self, comptime new_scope: anytype) *StandardLogger(new_scope) {
+            return @ptrCast(self);
+        }
+
+        pub fn spawn(self: *Self) void {
+            self.handle = std.Thread.spawn(.{}, Self.run, .{self}) catch @panic("could not spawn Logger");
+        }
+
+        pub fn deinit(self: *Self) void {
+            if (self.handle) |*handle| {
+                self.exit_sig.store(true, .seq_cst);
+                handle.join();
+            }
+            self.channel.close();
+            self.channel.deinit();
+            self.recycle_fba.deinit();
+        }
+
+        pub fn run(self: *Self) void {
+            while (!self.exit_sig.load(.seq_cst)) {
+                std.time.sleep(std.time.ns_per_ms * 5);
+                const messages = self.channel.drain() orelse {
+                    // channel is closed
+                    return;
+                };
+                defer self.channel.allocator.free(messages);
+
+                for (messages) |message| {
+                    const writer = std.io.getStdErr().writer();
+                    logfmt.writeLog(writer, message) catch @panic("logging failed");
+                }
+                self.recycle_fba.mux.lock();
+                self.recycle_fba.alloc_allocator.reset();
+                self.recycle_fba.mux.unlock();
+            }
+        }
+
+        pub fn log(self: *Self, level: Level, message: []const u8) void {
+            if (@intFromEnum(level) > @intFromEnum(self.max_level)) {
+                // noop
+                return;
+            }
+            const maybe_scope = blk: {
+                if (scope) |s| {
+                    break :blk @typeName(s);
+                } else {
+                    break :blk null;
+                }
+            };
+
+            const log_message = logfmt.createLogMessage(
+                &self.recycle_fba,
+                self.max_buffer,
+                level,
+                maybe_scope,
+                message,
+                null,
+                null,
+                null,
+            );
+            self.channel.send(log_message) catch @panic("could not send to channel");
```

Comment:
> ignores errors in the log* functions

> > I agree that's an improvement we can have in the new one.

Although this means if the logger runs into an error condition, it might be confounding to figure out why logs are not showing up (and no errors either) - an alternative may be to default to std.debug.err in this cases?

I won't do that now, and only ignore, but something to keep in mind if such a scenario pops up...

Reply:
> Actually it panics in more places than not. See

Sorry, I think I was confusing this with `std.debug.print`, which just ignores errors and silently fails to print. I should have checked the actual code.

I don't think it's necessary to ignore the errors. I see your point about silent failures being confusing. I'm just concerned about putting explicit panics in the code. I feel like panics should only be used very rarely in the specific situation I described here: https://github.com/Syndica/sig/pull/249#discussion_r1742294348. I don't think these panics fit those criteria. I personally would prefer to return an error instead of panic.

Reply:
I also agree that `fallback to std.debug.print` is probably the best option here. I have updated to ignoring error, but I i'll update in a bit to fallback to std.debug.print instead.